File: XamarinXamlPrev/ViewModel/ViewModelStopWatch.fs
namespace XamarinXamlPrev
open System
open System.Collections.Generic
open System.Linq
open System.Text
open Xamarin.Forms
open Moonmile.XFormsProvider
open System.Threading.Tasks
open System.Reflection

type ElapsedEventHandler = delegate of (obj * EventArgs) -> unit

type ViewModelStopWatch() =
    inherit ViewModelBase()

    let mutable startTime = DateTime.Now
    let mutable endTime  = DateTime.Now
    let mutable time:TimeSpan = new TimeSpan()
    let mutable task:Task = null
    let mutable taskflag = false


    /// リフレクションで iOS/Android の System.Timers.Timer を持ってくる
    /// 現在調節中
    let makeTimer (interval:int, hdr:ElapsedEventHandler) =
        let asm = Assembly.Load(AssemblyName("System"))
        let t = asm.GetType( "System.Timers.Timer" )
        let item = System.Activator.CreateInstance(t)

        t.GetRuntimeProperty("Enabled").SetValue( item, true )
        t.GetRuntimeProperty("AutoReset").SetValue( item, true )
        t.GetRuntimeProperty("Interval").SetValue( item, interval )
        t.GetRuntimeEvent("Elapsed").AddEventHandler( item, hdr )
        t.GetRuntimeMethod("Start", null).Invoke(item, null) |> ignore
        item


    member this.Time 
        with get() =  
            time <- endTime - startTime
            String.Format("{0:00}min {1:00}sec {2:000}", time.Minutes, time.Seconds, time.Milliseconds )
    member this.StartTime
        with get() = startTime.ToString()

    member this.StartCommand 
        with get() =
            let cmd = 
                new DelegateCommand(
                    (fun (o) -> true ),
                    (fun (o) -> 
                        startTime <- DateTime.Now 
                        endTime <- startTime
                        this.OnPropertyChanged("StartTime")
                        this.OnPropertyChanged("Time")
                        // makeTimer(2000, new ElapsedEventHandler(fun(s,e)-> 
                        //     System.Diagnostics.Debug.WriteLine("time {0}", endTime ))) |> ignore
                        // )))
                        ))
            cmd

    member this.StopCommand 
        with get() =
            let cmd = 
                new DelegateCommand(
                    (fun (o) -> true ),
                    (fun (o) -> 
                        taskflag <- false
                        endTime <- DateTime.Now 
                        this.OnPropertyChanged("Time")
                    ))    
            cmd

    member this.ResetCommand 
        with get() =
            let cmd = 
                new DelegateCommand(
                    (fun (o) -> true ),
                    (fun (o) -> 
                        taskflag <- false
                        startTime <- DateTime.Now
                        endTime <- DateTime.Now 
                        this.OnPropertyChanged("StartTime")
                        this.OnPropertyChanged("Time")
                    ))    
            cmd

File: CSharp/XamarinXamlPreviewCS/XamarinXamlPreviewCS/XamarinXamlPreviewCS/ViewModel/ViewModelStopWatch.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Reflection;
using System.Text;
using System.Threading.Tasks;
using System.Windows.Input;
using XamarinXamlPreviewCS.Libs;

namespace XamarinXamlPreviewCS.ViewModel
{
    public delegate void ElapsedEventHandler( object sender, EventArgs e);
    public class ViewModelStopWatch : Libs.ViewModelBase
    {
        protected DateTime startTime = DateTime.Now;
        protected DateTime endTime = DateTime.Now;
        protected TimeSpan time = new TimeSpan();

        /// リフレクションで iOS/Android の System.Timers.Timer を持ってくる
        /// 現在調節中
        public object makeTimer(int interval, ElapsedEventHandler hdr)
        {
            var asm = Assembly.Load(new AssemblyName("System"));
            var t = asm.GetType("System.Timers.Timer");
            var item = System.Activator.CreateInstance(t);

            t.GetRuntimeProperty("Enabled").SetValue(item, true);
            t.GetRuntimeProperty("AutoReset").SetValue( item, true );
            t.GetRuntimeProperty("Interval").SetValue( item, interval );
            t.GetRuntimeEvent("Elapsed").AddEventHandler(item, hdr);
            return item;
        }

        public string Time
        {
            get
            {
                time = endTime - startTime;
                return string.Format("{0:00}min {1:00}sec {2:000}", time.Minutes, time.Seconds, time.Milliseconds);
            }
        }
        public string StartTime
        {
            get { return startTime.ToString(); }
        }

        public ICommand StartCommand
        {
            get
            {
                var cmd = new DelegateCommand(
                    (o) => true,
                    (o) => {
                        startTime = DateTime.Now;
                        endTime = startTime;
                        this.OnPropertyChanged("StartTime");
                        this.OnPropertyChanged("Time");
                    });
                return cmd;
            }
        }
        public ICommand StopCommand
        {
            get
            {
                var cmd = new DelegateCommand(
                    (o) => true,
                    (o) =>
                    {
                        endTime = DateTime.Now;
                        this.OnPropertyChanged("Time");
                    });
                return cmd;
            }
        }
        public ICommand ResetCommand
        {
            get
            {
                var cmd = new DelegateCommand(
                    (o) => true,
                    (o) =>
                    {
                        startTime = DateTime.Now;
                        endTime = startTime;
                        this.OnPropertyChanged("StartTime");
                        this.OnPropertyChanged("Time");
                    });
                return cmd;
            }
        }
    }
}
